### PR TITLE
Add new authenticated encryption algorithm

### PR DESCRIPTION
**Pull request by AnastasiaKubova**
Things to consider before you submit the PR:

* [x] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [x] Did you check UI tests on the sample app? They are not executed on CI.
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

- Add new authenticated encryption algorithm (EtM)
- Add usage of new authenticated encryption algorithm

## Related PRs or issues

[AB#86674](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/86674)
[AB#86677](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/86677)
[AB#86675](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/86675)

### CODE REVIEW

File: AppCenter/AppCenter/Internals/Util/MSACEncrypter.m [hunk 1]
```diff
@@ -220,13 +279,60 @@ + (NSData *)generateInitializationVector {
...
++ (NSData *)getMetadataStringWithKeyTag:(NSString *)keyTag encryptionAlgorithmName:(NSString *)encryptionAlgorithmName {
 
   // Format is {key tag}/{algorithm}/{cipher mode}/{padding mode}/{key length}
   NSArray *metadata =
-      @[ keyTag, kMSACEncryptionAlgorithmName, kMSACEncryptionCipherMode, kMSACEncryptionPaddingMode, @(kMSACEncryptionKeySize) ];
+      @[ keyTag, encryptionAlgorithmName, kMSACEncryptionCipherMode, kMSACEncryptionPaddingMode, @(kMSACEncryptionKeySize) ];
   NSString *metadataString = [metadata componentsJoinedByString:kMSACEncryptionMetadataInternalSeparator];
-  return [metadataString dataUsingEncoding:NSUTF8StringEncoding];
+  NSData *encodingData = [metadataString dataUsingEncoding:NSUTF8StringEncoding];
+  return encodingData;
```
Comment (DmitriyKirakosyan) [inline AppCenter/AppCenter/Internals/Util/MSACEncrypter.m:289]: let's just `return [metadataString dataUsingEncoding:NSUTF8StringEncoding];`

File: AppCenter/AppCenter/Internals/Util/MSACEncrypter.m [hunk 2]
```diff
@@ -220,13 +279,60 @@ + (NSData *)generateInitializationVector {
...
+}
+
++ (NSData *)getMetadataStringWithKeyTag:(NSString *)keyTag {
+  return [MSACEncrypter getMetadataStringWithKeyTag:keyTag encryptionAlgorithmName:kMSACEncryptionAlgorithmName];
+}
+
+- (NSData *_Nonnull)getMacBytes:(NSData *_Nonnull)key keySize:(NSData *_Nonnull)cipherText {
```
Comment (DmitriyKirakosyan) [inline AppCenter/AppCenter/Internals/Util/MSACEncrypter.m:296]: why do we need `_Nonnull`s here?

File: AppCenter/AppCenter/Internals/Util/MSACEncrypter.m [hunk 3]
```diff
@@ -220,13 +279,60 @@ + (NSData *)generateInitializationVector {
...
+  NSData *hMAC = [[NSData alloc] initWithBytes:cHMAC length:sizeof(cHMAC)];
+  return hMAC;
+}
+
+- (NSData *_Nonnull)getSubkey:(NSData *_Nonnull)secretKey outputSize:(int)outputDataLength {
+
+  // Check output data length.
+  if (outputDataLength < 1) {
+    NSException *outputDataLengthException = [NSException exceptionWithName:@"Invalid output length data."
+                                                                     reason:@"Output data length must be at more than zero."
```
Comment (DmitriyKirakosyan) [inline AppCenter/AppCenter/Internals/Util/MSACEncrypter.m:308]:
```suggestion
                                                                     reason:@"Output data length must be greater than zero."
```

File: AppCenter/AppCenter/Internals/Util/MSACEncrypter.m [hunk 4]
```diff
@@ -220,13 +279,60 @@ + (NSData *)generateInitializationVector {
...
+  CCHmac(kCCHmacAlgSHA256, key.bytes, key.length, cipherText.bytes, cipherText.length, cHMAC);
+  NSData *hMAC = [[NSData alloc] initWithBytes:cHMAC length:sizeof(cHMAC)];
+  return hMAC;
+}
+
+- (NSData *_Nonnull)getSubkey:(NSData *_Nonnull)secretKey outputSize:(int)outputDataLength {
+
+  // Check output data length.
+  if (outputDataLength < 1) {
+    NSException *outputDataLengthException = [NSException exceptionWithName:@"Invalid output length data."
```
Comment (DmitriyKirakosyan) [inline AppCenter/AppCenter/Internals/Util/MSACEncrypter.m:307]:
```suggestion
    NSException *outputDataLengthException = [NSException exceptionWithName:@"Invalid output data length."
```

File: AppCenter/AppCenter/Internals/Util/MSACEncrypter.m [hunk 5]
```diff
@@ -220,13 +279,60 @@ + (NSData *)generateInitializationVector {
...
+
+  // Calculate iterations.
+  int iterations = (int)(ceil((double)outputDataLength / (double)CC_SHA256_DIGEST_LENGTH));
+
+  // Prepare data.
+  NSData *mixin = [NSData data];
+  NSMutableData *results = [NSMutableData data];
+
+  // Calculate subkey.
+  for (int i = 0; i < (iterations); i++) {
```
Comment (DmitriyKirakosyan) [inline AppCenter/AppCenter/Internals/Util/MSACEncrypter.m:321]:
```suggestion
  for (int i = 0; i < iterations; i++) {
```

File: AppCenter/AppCenter/Internals/Util/MSACEncrypter.m [hunk 6]
```diff
@@ -220,13 +279,60 @@ + (NSData *)generateInitializationVector {
...
+                                                                     reason:@"Output data length must be at more than zero."
+                                                                   userInfo:nil];
+    @throw outputDataLengthException;
+  }
+
+  // Calculate iterations.
+  int iterations = (int)(ceil((double)outputDataLength / (double)CC_SHA256_DIGEST_LENGTH));
+
+  // Prepare data.
+  NSData *mixin = [NSData data];
```
Comment (DmitriyKirakosyan) [inline AppCenter/AppCenter/Internals/Util/MSACEncrypter.m:317]: "mixin"?

File: AppCenter/AppCenter/Internals/Util/MSACEncrypter.m [hunk 7]
```diff
@@ -60,37 +76,80 @@ - (NSString *_Nullable)decryptString:(NSString *)string {
...
-    // Extract key from metadata.
-    NSString *keyTag = [metadata componentsSeparatedByString:kMSACEncryptionMetadataInternalSeparator][0];
-
-    // Metadata, separator, and initialization vector.
-    size_t cipherTextPrefixLength = metadataLocation + 1 + kCCBlockSizeAES128;
-    NSRange cipherTextRange = NSMakeRange(cipherTextPrefixLength, [data length] - cipherTextPrefixLength);
     NSRange ivRange = NSMakeRange(metadataLocation + 1, kCCBlockSizeAES128);
-    initializationVector = [data subdataWithRange:ivRange];
-    cipherText = [data subdataWithRange:cipherTextRange];
-    key = [self getKeyWithKeyTag:keyTag];
```
Comment (DmitriyKirakosyan) [inline AppCenter/AppCenter/Internals/Util/MSACEncrypter.m:86]: Why did you remove this line. `[self getKeyWithKeyTag:keyTag]` is called multiple times below. Let's save it to var.

**Review by DmitriyKirakosyan — CHANGES_REQUESTED** [review]

File: AppCenter/AppCenter/Internals/Util/MSACEncrypter.m (hunk 7 above)
Comment (AnastasiaKubova) [reply AppCenter/AppCenter/Internals/Util/MSACEncrypter.m:86]: Nope, it is called only once.

File: AppCenter/AppCenter/Internals/Util/MSACEncrypter.m [hunk 8]
```diff
@@ -220,13 +280,68 @@ + (NSData *)generateInitializationVector {
...
 }
 
++ (NSData *)getMetadataStringWithKeyTag:(NSString *)keyTag {
+  return [MSACEncrypter getMetadataStringWithKeyTag:keyTag encryptionAlgorithmName:kMSACEncryptionAlgorithmName];
+}
+
+- (NSData *)getMacBytes:(NSData *_Nonnull)key keySize:(NSData *_Nonnull)cipherText {
+  unsigned char cHMAC[CC_SHA256_DIGEST_LENGTH];
+  CCHmac(kCCHmacAlgSHA256, key.bytes, key.length, cipherText.bytes, cipherText.length, cHMAC);
+  NSData *hMAC = [[NSData alloc] initWithBytes:cHMAC length:sizeof(cHMAC)];
```
Comment (DmitriyKirakosyan) [inline AppCenter/AppCenter/Internals/Util/MSACEncrypter.m:299]: Let's just return this value. I think we don't need an extra var here.

File: AppCenter/AppCenterTests/MSACEncrypterTests.m [hunk 9]
```diff
@@ -362,12 +362,103 @@ - (void)testDecryptWithExpiredKey {
...
+  NSString *metadata = [NSString stringWithFormat:@"%@/AES/HmacSHA256/CBC/PKCS7/32", keyTag];
+  NSString *keyMetadataString = [NSString stringWithFormat:@"%@/%@", keyTag, expirationIso];
+  [MSAC_APP_CENTER_USER_DEFAULTS setObject:keyMetadataString forKey:kMSACEncryptionKeyMetadataKey];
+
+  // Save key to the Keychain.
+  NSString *currentKey = [self generateTestEncryptionKey];
+  [MSACMockKeychainUtil storeString:currentKey forKey:keyTag];
+  NSData *key = [[NSData alloc] initWithBase64EncodedString:currentKey options:0];
+
+  // Prepare encrypted test with old encryption.
```
Comment (DmitriyKirakosyan) [inline AppCenter/AppCenterTests/MSACEncrypterTests.m:382]:
```suggestion
  // Prepare encrypted text with old encryption.
```

File: AppCenter/AppCenterTests/MSACEncrypterTests.m [hunk 10]
```diff
@@ -362,12 +362,103 @@ - (void)testDecryptWithExpiredKey {
...
+  NSString *oldMetadata = [NSString stringWithFormat:@"%@/AES/CBC/PKCS7/32", keyTag];
+  NSString *keyMetadataString = [NSString stringWithFormat:@"%@/%@", keyTag, expirationIso];
+  [MSAC_APP_CENTER_USER_DEFAULTS setObject:keyMetadataString forKey:kMSACEncryptionKeyMetadataKey];
+
+  // Save key to the Keychain.
+  NSString *currentKey = [self generateTestEncryptionKey];
+  [MSACMockKeychainUtil storeString:currentKey forKey:keyTag];
+  NSData *key = [[NSData alloc] initWithBase64EncodedString:currentKey options:0];
+
+  // Prepare encrypted test with old encryption.
```
Comment (DmitriyKirakosyan) [inline AppCenter/AppCenterTests/MSACEncrypterTests.m:413]:
```suggestion
  // Prepare encrypted text with old encryption.
```

**Review by DmitriyKirakosyan — APPROVED** [review]
